lib/barlix/ean13.ex

```diff
@@ -0,0 +1,153 @@
+defmodule Barlix.EAN13 do
+  require Integer
+  use Bitwise, only_operators: true
+  import Integer, only: [mod: 2]
+
+  @moduledoc """
+  Implements [EAN13](https://en.wikipedia.org/wiki/International_Article_Number).
+  """
+
+  @doc """
+  Encodes the given value using EAN13. The given code is validated first.
+
+  ## Examples:
+
+    iex> Barlix.EAN13.encode("5449000096241")
+    {:ok, {:D1, [
+    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0
+    ]}}
+
+
+    iex> Barlix.EAN13.encode("5901234123450")
+    {:error, "validation failed"}
+
+  """
+  @spec encode(String.t()) :: {:error, String.t()} | {:ok, Barlix.code()}
+  def encode(value) when is_binary(value) and byte_size(value) == 13 do
+    case validate(value) do
+      :ok -> get_code(value)
+      e -> e
+    end
+  end
+
+  @doc """
+  Accepts the same arguments as `encode/1` but raises on error.
+  """
+  @spec encode!(String.t() | charlist) :: Barlix.code() | no_return
+  def encode!(value) when is_binary(value) and byte_size(value) == 13 do
+    case encode(value) do
+      {:ok, code} -> code
+      {:error, error} -> raise Barlix.Error, error
+    end
+  end
+
+  @doc """
+  Validate an EAN13 code.
+
+  ## Examples:
+
+    iex> Barlix.EAN13.validate("5449000096241")
+    :ok
+
+    iex> Barlix.EAN13.validate("5901234123450")
+    {:error, "validation failed"}
+  """
+  @spec validate(String.t() | charlist) :: :ok | {:error, String.t()}
+  def validate(<<value::binary-size(12), c::binary-size(1)>>) do
+    n =
+      value
+      |> String.split("", trim: true)
+      |> Enum.map(&String.to_integer/1)
+      |> Enum.with_index()
+      |> Enum.reduce(0, fn
+        {v, i}, acc when Integer.is_even(i) -> acc + v
+        {v, i}, acc when Integer.is_odd(i) -> acc + 3 * v
+      end)
+
+    checkdigit = mod(10 - mod(n, 10), 10) |> Integer.to_string()
+    if checkdigit == c, do: :ok, else: {:error, "validation failed"}
+  end
+
+  @spec get_code(String.t()) :: {:ok, Barlix.code()}
+  def get_code(<<p::binary-size(1), l::binary-size(6), r::binary-size(6)>>) do
+    prefix = String.to_integer(p)
+    encoding = ctbl(prefix)
+
+    left =
+      l
+      |> String.split("", trim: true)
+      |> Enum.map(&String.to_integer/1)
```

Comment:
The conversion of string to list of integers is being done at multiple places. This is preferable done at one place, probably the result of validate call and deal with integer list everywhere.

Reply:
Fixed it…